### PR TITLE
CI: Some Azure Pipelines cleanups

### PR DESCRIPTION
**Pull request by vtbassmatt**
Hey all, I'm a PM on Azure Pipelines. I noticed you had some comments in your `azure-pipelines.yml` regarding conditional steps and possibly adding a Linux flavor. I've taken the liberty of addressing those issues.

One thing I didn't completely finish: getting Linux to publish test results. The problem is that there are directories the publish task can't read already sitting in `/tmp`. I think the right solution is to use `mktemp -d` and then use that directory in the test scripts. But that felt like a pretty invasive change that you'd want to discuss before I just powered through and did it :)

Also, the contents of `azure-Linux-35.yaml` are identical to macOS right now, because I wasn't sure what needed to differ between those environments.

### CODE REVIEW

**Comment by datapythonista** [comment]
Thanks for the clean up @vtbassmatt. I don't know much about the Windows part, but lgtm. @TomAugspurger ?

@vtbassmatt unrelated to this PR, I'm working on moving the linting from travis to azure-pipelines in #22854. If you have feedback about it, that would be great.

File: ci/azure-Linux-35.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: pandas
```
Comment (alimcmaster1) [inline ci/azure-Linux-35.yaml]: Minor issue but should "Linux" be all lower case in the file name to be consistent with the windows yaml files?

**Comment by vtbassmatt** [comment]
@datapythonista I will definitely take a look. Probably Monday at this point.

**Comment by TomAugspurger** [comment]
I haven't looked too deeply, but the logs in https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=1216&view=logs reference numpy 1.10. We recently bumped our minimum to numpy 1.12.

**Comment by datapythonista** [comment]
@vtbassmatt can you rebase?

**Comment by vtbassmatt** [comment]
Rebased. Hopefully I didn't break anything -- I still have a shadow pipeline definition set up, so I'll know shortly.

**Comment by vtbassmatt** [comment]
[Mine looks clean](https://mattc-demo.visualstudio.com/python-ci-testing/_build/results?buildId=612&view=logs).

File: azure-pipelines.yml [hunk 2]
```diff
@@ -1,25 +1,17 @@
...
     vmImage: xcode9-macos10.13
-- template: ci/azure/linux.yml
-  parameters:
-    name: Linux
-    vmImage: ubuntu-16.04
 
-# Windows Python 2.7 needs VC 9.0 installed, and not sure
-# how to make that a conditional task, so for now these are
-# separate templates as well
+# - template: ci/azure/macos-linux.yml
```
Comment (TomAugspurger) [inline azure-pipelines.yml]: Is this commented out right now, or is being substituted?

**Review by TomAugspurger — COMMENTED** [review]
I think there are some issues here. I believe we lost the following builds (which were added after you started this PR @vtbassmatt)

- azure-27-compat
- azure-36-locale_slow
- azure-37-locale

I don't see a linux build in https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=2458&view=logs

So, could we add an `ENV_FILE` to the parameters passed to the `macos-linux` file, and then use those parameters in `azure-pipelines.yml` to have the following builds

- macOS, ci/azure-macos-35.yaml
- linux, azure-27-compat
- linux, azure-36-locale_slow
- linux, azure-37-locale

File: azure-pipelines.yml (hunk 2 above)
Comment (vtbassmatt) [reply azure-pipelines.yml]: Whoops, I really screwed this rebase up. Will fix it tomorrow.

**Comment by TomAugspurger** [comment]
No worries, I think it's our fault for letting this sit unmerged for too
long :) I believe the new linux builds were added after you opened the PR.

**Comment by vtbassmatt** [comment]
Let's see if that rebase was a little better...

**Comment by vtbassmatt** [comment]
I'm sorry, I wasn't watching this closely enough and missed the build failure.

**Comment by vtbassmatt** [comment]
It looks like the Mac agent freaked out on that last run. It was working in the prior run and I didn't change anything, plus it worked on my private copy of the pipeline. I think this is back in working order.

**Comment by jreback** [comment]
is this PR still relevant? @datapythonista

**Comment by datapythonista** [comment]
yes, that's something we'd like to have, as among other things it avoids repetitive code in the CI config.

@vtbassmatt can you merge master and fix the conflicts? And as I think you're addressing different things here, do you think it makes sense to split this PR, so we can merge the parts that are ready?

**Comment by vtbassmatt** [comment]
I'll take a look tomorrow. I had already resolved a similar rebase before, and it took a bit of finesse to make sure I didn't break anything.

**Comment by vtbassmatt** [comment]
@datapythonista pending the outcome of the Azure build, I think this is ready to merge. Can we please merge it soon so I don't have to rebase again? 😋

**Review by datapythonista — APPROVED** [review]
Thanks @vtbassmatt, nice job, great to have this simplified.

lgtm, but I'm a bit lost reviewing the Windows part, @TomAugspurger can you take a look please?

**Comment by datapythonista** [comment]
Btw @vtbassmatt I think the problem we've got with Azure not reporting the end of the build to GitHub happens in all the PRs now, including this one. Not sure if it happens to other projects too, but probably worth reporting it again.

**Comment by vtbassmatt** [comment]
We think we've spotted the bug - a missing null check where the build errors without a message. We're preparing a hotfix now.

**Comment by datapythonista** [comment]
That's great, thanks @vtbassmatt, and thanks for keeping us updated.

@TomAugspurger, now that #24092 has been merged, if you can take a look at this PR, I think it should be ready to be merged too.

**Comment by datapythonista** [comment]
@TomAugspurger if you can take a look, and merge if you're happy

**Comment by TomAugspurger** [comment]
I probably won't have time to look today, so I'd say go ahead and merge if
you're comfortable with the changes.

**Comment by datapythonista** [comment]
Double checked the changes, and the build, and everything looks all right. Merging.

Thanks @vtbassmatt

**Pull request merged**